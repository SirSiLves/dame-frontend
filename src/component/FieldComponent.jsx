import React, {Component} from 'react';
import StoneComponent from "./StoneComponent";
import "../stylesheet/field.css";
import crown from "../crown.png";


class FieldComponent extends Component {
    constructor(props) {
        super(props);
        this.handleStoneClick = this.handleStoneClick.bind(this);
    }

    render() {

        return (
            <div className={this.getStyles()} onClick={this.getOnFieldClick()}>
                {this.setFieldNumber()}
                {this.showMove()}
                <StoneComponent stone={this.props.field.stone} onClick={this.handleStoneClick}/>
            </div>
        )
    }

    //animation handling after a move was triggered
    showMove() {
        if (this.props.moveHistory !== null && this.props.moveHistory !== undefined) {

            let moveEntriesCount = Object.keys(this.props.moveHistory).length;
            let lastMoveEntry = this.props.moveHistory[moveEntriesCount - 1];

            if (moveEntriesCount > 0) {

                if (lastMoveEntry.lastHoppedField.field != null) {
                    //animation killed stone
                    if (this.props.field.fieldNumber === lastMoveEntry.lastHoppedField.field.fieldNumber) {
                        let styleClass = "stone stone-killed ";
                        lastMoveEntry.lastTargetField.stone.stoneColor.toLowerCase() === "black" ? styleClass += "white" : styleClass += "black";
                        if(lastMoveEntry.lastHoppedField.queenBefore){
                            return <div className={styleClass}>
                                <img src={crown} className="queen" alt={"This is a queen"} height={40} width={40}/>
                            </div>
                        }
                        return <div className={styleClass}/>
                    }

                    //go through the history, if double (or more) hop was done
                    for (let i = moveEntriesCount - 1; i > 0; i--) {
                        if (this.props.moveHistory[i].lastCurrentField.field.fieldNumber === this.props.moveHistory[i - 1].lastTargetField.field.fieldNumber) {
                            //show stopover
                            if (this.props.field.fieldNumber === this.props.moveHistory[i - 1].lastTargetField.field.fieldNumber) {
                                if (lastMoveEntry.lastTargetField.field.stone != null) {
                                    let styleClass = "stone stone-hide " + lastMoveEntry.lastTargetField.field.stone.stoneColor.toLowerCase();

                                    if(this.props.moveHistory[i - 1].lastTargetField.queenBefore){
                                        return <div className={styleClass}>
                                            <img src={crown} className="queen" alt={"This is a queen"} height={40} width={40}/>
                                        </div>
                                    }
                                    return <div className={styleClass}/>
                                }
                            }

                            //animation killed stone
                            if (this.props.moveHistory[i - 1].lastHoppedField.field != null &&
                                this.props.field.fieldNumber === this.props.moveHistory[i - 1].lastHoppedField.field.fieldNumber) {
                                let styleClass = "stone stone-killed ";
                                lastMoveEntry.lastTargetField.stone.stoneColor.toLowerCase() === "black" ? styleClass += "white" : styleClass += "black";

                                if(this.props.moveHistory[i - 1].lastHoppedField.queenBefore){
                                    return <div className={styleClass}>
                                        <img src={crown} className="queen" alt={"This is a queen"} height={40} width={40}/>
                                    </div>
                                }
                                return <div className={styleClass}/>
                            }
                        } else {
                            break;
                        }
                    }
                }
            }
        }
    }

    //build the className for the field
    getStyles() {
        let styles = "";
        styles += this.props.field.fieldColor.toLowerCase();
        styles += " ";
        styles += "field";
        return styles;
    }

    //choose the clicked field
    getOnFieldClick() {
        if (this.props.field.stone) {
            return null
        } else {
            return this.props.onFieldClick
        }
    }

    //shows the field numbers in the debug mode
    setFieldNumber() {
        if (process.env.REACT_APP_DEBUG_MODE === "enabled") return this.props.field.fieldNumber
    }

    //passing up the data to the parent component
    handleStoneClick() {
        this.props.currentFieldCallbackFieldLine(this.props.field);
    }
}

export default FieldComponent
