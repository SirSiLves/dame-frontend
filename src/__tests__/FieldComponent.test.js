import React from 'react';
import ReactDOM from 'react-dom';
import {shallow} from 'enzyme';
import FieldComponent from "../component/FieldComponent";

it('renders FieldComponent without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<FieldComponent key={1} field={field} />, div);
});

describe('FieldComponent Component', () => {
    it('should render correctly in "debug" mode', () => {
        const component = shallow(<FieldComponent key={1} field={field} debug/>);

        expect(component).toMatchSnapshot();
    });
});

const field =
    {
        fieldColor: "SANDYBROWN",
        fieldNumber: 0,
        stone: null
    };