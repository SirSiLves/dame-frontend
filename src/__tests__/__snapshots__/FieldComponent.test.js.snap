// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldComponent Component should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;
